File: ios/PlaytomicFoundation/executor/Executor.swift
//
//  Executor.swift
//  PlaytomicFoundation
//
//  Created by Manuel González Villegas on 28/11/2019.
//  Copyright © 2019 Playtomic. All rights reserved.
//

import Foundation

public class Executor {
    public class CancellableTask {
        public enum Status {
            case waiting, completed, canceled
        }

        public private(set) var status: Status = .waiting
        var task: DispatchWorkItem!
        private let job: () -> Void

        init(job: @escaping () -> Void) {
            self.job = job
            self.task = DispatchWorkItem { self.run() }
        }

        public func run() {
            if status == .waiting {
                job()
                status = .completed
                task = nil
            }
        }

        public func cancel() {
            if status == .waiting {
                task.cancel()
                task = nil
                status = .canceled
            }
        }
    }

    @discardableResult
    public static func execute(after: TimeInterval = 0, inBackground: Bool, job: @escaping () -> Void) -> CancellableTask? {
        if after > 0 || inBackground || !Thread.isMainThread {
            let cancellableTask = CancellableTask(job: job)
            let queue = inBackground ? DispatchQueue.global(qos: .background) : DispatchQueue.main
            queue.asyncAfter(deadline: .now() + after, execute: cancellableTask.task)
            return cancellableTask
        } else {
            job()
            return nil
        }
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/executor/Executor.kt
package com.playtomic.foundation.executor

import android.os.Handler
import android.os.Looper
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

object PlaytomicExecutor {
    private const val DEFAULT_THREAD_POOL_SIZE = 4
    private val executorService: ExecutorService = Executors.newFixedThreadPool(DEFAULT_THREAD_POOL_SIZE)

    fun execute(runnable: () -> Unit) {
        executorService.execute(runnable)
    }
}

class Executor {
    class CancellableTask(private val job: () -> Unit, private val handler: Handler) : Runnable {
        enum class Status {
            waiting, completed, canceled
        }
        var status: Status = Status.waiting
            private set

        override fun run() {
            if (status == Status.waiting) {
                job()
                status = Status.completed
            }
        }

        fun cancel() {
            if (status == Status.waiting) {
                handler.removeCallbacks(this)
                status = Status.canceled
            }
        }
    }

    companion object {
        private var mainHandler = Handler(Looper.getMainLooper())

        var backgroundExecutor: (() -> Unit) -> Unit = {
            PlaytomicExecutor.execute(it)
        }

        var mainExecutor: (() -> Unit) -> Unit = {
            if (isMainThread) {
                it()
            } else {
                mainHandler.post(it)
            }
        }

        private val isMainThread: Boolean
            get() = Looper.myLooper() == Looper.getMainLooper()

        fun execute(after: Double = 0.0, inBackground: Boolean, job: () -> Unit): CancellableTask? {
            return if (after > 0 || inBackground || !isMainThread) {
                val executor = if (inBackground) backgroundExecutor else mainExecutor
                val task = CancellableTask({ executor(job) }, mainHandler)
                mainHandler.postDelayed(task, (after * 1000).toLong())
                task
            } else {
                job()
                null
            }
        }
    }
}
